File: nexago_app/lib/features/arenas/domain/arena_amenities.dart
/// Comodidades da arena (`arenas/{id}.amenities` no Firestore).
class ArenaAmenities {
  const ArenaAmenities({
    this.parking = false,
    this.lockerRoom = false,
    this.coveredCourt = false,
    this.bar = false,
    this.racketRental = false,
    this.hasAccessibleCourt = false,
    this.hasAccessibleBathroom = false,
    this.hasPcdParking = false,
  });

  final bool parking;
  final bool lockerRoom;
  final bool coveredCourt;
  final bool bar;
  final bool racketRental;
  final bool hasAccessibleCourt;
  final bool hasAccessibleBathroom;
  final bool hasPcdParking;

  static const ArenaAmenities empty = ArenaAmenities();

  bool get hasAny =>
      parking ||
      lockerRoom ||
      coveredCourt ||
      bar ||
      racketRental ||
      hasAccessibleCourt ||
      hasAccessibleBathroom ||
      hasPcdParking;

  /// Requisitos de filtro: cada `true` exige que a arena tenha a comodidade.
  bool matchesRequirements(ArenaAmenities required) {
    if (required.parking && !parking) return false;
    if (required.lockerRoom && !lockerRoom) return false;
    if (required.coveredCourt && !coveredCourt) return false;
    if (required.bar && !bar) return false;
    if (required.racketRental && !racketRental) return false;
    if (required.hasAccessibleCourt && !hasAccessibleCourt) return false;
    if (required.hasAccessibleBathroom && !hasAccessibleBathroom) {
      return false;
    }
    if (required.hasPcdParking && !hasPcdParking) return false;
    return true;
  }

  factory ArenaAmenities.fromMap(Map<String, dynamic>? map) {
    if (map == null || map.isEmpty) return empty;

    bool read(String key, List<String> aliases) {
      for (final k in [key, ...aliases]) {
        final v = map[k];
        if (v is bool) return v;
      }
      return false;
    }

    return ArenaAmenities(
      parking: read('parking', ['estacionamento', 'hasParking']),
      lockerRoom: read('lockerRoom', ['locker_room', 'vestiario', 'vestiário']),
      coveredCourt: read('coveredCourt', [
        'covered_court',
        'coberta',
        'indoor',
        'covered',
      ]),
      bar: read('bar', ['hasBar']),
      racketRental: read('racketRental', [
        'racket_rental',
        'aluguel_raquetes',
        'racketRental',
      ]),
      hasAccessibleCourt: read('hasAccessibleCourt', [
        'quadra_acessivel',
        'accessible_court',
      ]),
      hasAccessibleBathroom: read('hasAccessibleBathroom', [
        'banheiro_acessivel',
        'accessible_bathroom',
      ]),
      hasPcdParking: read('hasPcdParking', [
        'vaga_pcd',
        'pcd_parking',
      ]),
    );
  }

  Map<String, dynamic> toFirestoreMap() => {
        'parking': parking,
        'lockerRoom': lockerRoom,
        'coveredCourt': coveredCourt,
        'bar': bar,
        'racketRental': racketRental,
        'hasAccessibleCourt': hasAccessibleCourt,
        'hasAccessibleBathroom': hasAccessibleBathroom,
        'hasPcdParking': hasPcdParking,
      };

  ArenaAmenities copyWith({
    bool? parking,
    bool? lockerRoom,
    bool? coveredCourt,
    bool? bar,
    bool? racketRental,
    bool? hasAccessibleCourt,
    bool? hasAccessibleBathroom,
    bool? hasPcdParking,
  }) {
    return ArenaAmenities(
      parking: parking ?? this.parking,
      lockerRoom: lockerRoom ?? this.lockerRoom,
      coveredCourt: coveredCourt ?? this.coveredCourt,
      bar: bar ?? this.bar,
      racketRental: racketRental ?? this.racketRental,
      hasAccessibleCourt: hasAccessibleCourt ?? this.hasAccessibleCourt,
      hasAccessibleBathroom:
          hasAccessibleBathroom ?? this.hasAccessibleBathroom,
      hasPcdParking: hasPcdParking ?? this.hasPcdParking,
    );
  }

  @override
  bool operator ==(Object other) {
    return other is ArenaAmenities &&
        other.parking == parking &&
        other.lockerRoom == lockerRoom &&
        other.coveredCourt == coveredCourt &&
        other.bar == bar &&
        other.racketRental == racketRental &&
        other.hasAccessibleCourt == hasAccessibleCourt &&
        other.hasAccessibleBathroom == hasAccessibleBathroom &&
        other.hasPcdParking == hasPcdParking;
  }

  @override
  int get hashCode => Object.hash(
        parking,
        lockerRoom,
        coveredCourt,
        bar,
        racketRental,
        hasAccessibleCourt,
        hasAccessibleBathroom,
        hasPcdParking,
      );
}

File: frontend/shared/arena-discovery/arena-amenities.ts
/** Comodidades da arena (`arenas/{id}.amenities` no Firestore). Paridade com `ArenaAmenities` (Flutter). */
export interface ArenaAmenities {
  parking: boolean;
  lockerRoom: boolean;
  coveredCourt: boolean;
  bar: boolean;
  racketRental: boolean;
  hasAccessibleCourt: boolean;
  hasAccessibleBathroom: boolean;
  hasPcdParking: boolean;
}

export const ARENA_AMENITIES_EMPTY: ArenaAmenities = {
  parking: false,
  lockerRoom: false,
  coveredCourt: false,
  bar: false,
  racketRental: false,
  hasAccessibleCourt: false,
  hasAccessibleBathroom: false,
  hasPcdParking: false,
};

export function amenitiesHasAny(a: ArenaAmenities): boolean {
  return (
    a.parking ||
    a.lockerRoom ||
    a.coveredCourt ||
    a.bar ||
    a.racketRental ||
    a.hasAccessibleCourt ||
    a.hasAccessibleBathroom ||
    a.hasPcdParking
  );
}

/** Cada `true` em `required` exige que `actual` também tenha a comodidade. */
export function amenitiesMatchesRequirements(
  actual: ArenaAmenities,
  required: ArenaAmenities,
): boolean {
  if (required.parking && !actual.parking) return false;
  if (required.lockerRoom && !actual.lockerRoom) return false;
  if (required.coveredCourt && !actual.coveredCourt) return false;
  if (required.bar && !actual.bar) return false;
  if (required.racketRental && !actual.racketRental) return false;
  if (required.hasAccessibleCourt && !actual.hasAccessibleCourt) return false;
  if (required.hasAccessibleBathroom && !actual.hasAccessibleBathroom) return false;
  if (required.hasPcdParking && !actual.hasPcdParking) return false;
  return true;
}

function readBool(map: Record<string, unknown>, key: string, aliases: readonly string[]): boolean {
  for (const k of [key, ...aliases]) {
    const v = map[k];
    if (typeof v === 'boolean') return v;
  }
  return false;
}

export function amenitiesFromFirestore(raw: unknown): ArenaAmenities {
  if (raw == null || typeof raw !== 'object') {
    return ARENA_AMENITIES_EMPTY;
  }
  const map = raw as Record<string, unknown>;
  return {
    parking: readBool(map, 'parking', ['estacionamento', 'hasParking']),
    lockerRoom: readBool(map, 'lockerRoom', ['locker_room', 'vestiario', 'vestiário']),
    coveredCourt: readBool(map, 'coveredCourt', ['covered_court', 'coberta', 'indoor', 'covered']),
    bar: readBool(map, 'bar', ['hasBar']),
    racketRental: readBool(map, 'racketRental', ['racket_rental', 'aluguel_raquetes']),
    hasAccessibleCourt: readBool(map, 'hasAccessibleCourt', ['quadra_acessivel', 'accessible_court']),
    hasAccessibleBathroom: readBool(map, 'hasAccessibleBathroom', ['banheiro_acessivel', 'accessible_bathroom']),
    hasPcdParking: readBool(map, 'hasPcdParking', ['vaga_pcd', 'pcd_parking']),
  };
}
